### PR TITLE
RTE: enable per-node-group tuning

### PR DESCRIPTION
**Pull request by fromanirh**
Introduce per-node-group config (NodeGroupConfig) object, in order to enable fine-tuning of RTE.

This is needed to enable cluster-admins to explore the proper tradeoff between cluster load and scheduling precision.

We will also need to have some slack to explore the best default to set, and we need some scalability knobs for the scale benchmarks (due "soon").

The new Config object means we need to deprecate the `DisablePodsFingerprinting` bool we added in the 4.11 cycle.
This also plays nice with the ongoing deprecation of the bool flags in the k8s APIs.

### CODE REVIEW

**Comment by fromanirh** [comment]
/hold
missing controller tests. We don't need e2e tests, but controller tests we definitely need.

**Comment by fromanirh** [comment]
/hold cancel
tests added

File: bundle/manifests/numaresources-operator.clusterserviceversion.yaml [hunk 1]
```diff
@@ -65,6 +65,11 @@ spec:
       - description: Group of Nodes to enable RTE on
         displayName: Group of nodes to enable RTE on
         path: nodeGroups
+      - description: Set the topology info refresh period. Use explicit 0 to disable.
```
Comment (swatisehgal) [inline bundle/manifests/numaresources-operator.clusterserviceversion.yaml:68]: Why is only the refresh period here and not `infoRefreshMode`? The refresh period is linked to `infoRefreshMode` and depends on whether `Periodic` or `Evented` so was expecting both here.

File: pkg/objectstate/merge/merge.go [hunk 2]
```diff
@@ -96,3 +98,19 @@ func Labels(current, updated client.Object) client.Object {
...
+
+	if updated.PodsFingerprinting != nil {
+		conf.PodsFingerprinting = updated.PodsFingerprinting
+	}
+	if updated.InfoRefreshPeriod != nil {
+		conf.InfoRefreshPeriod = updated.InfoRefreshPeriod
+	}
+	if updated.InfoRefreshMode != nil {
+		conf.InfoRefreshMode = updated.InfoRefreshMode
+	}
```
Comment (swatisehgal) [inline pkg/objectstate/merge/merge.go:114]: What is the value of `InfoRefreshPeriod` if `Evented` mode is specified? IIUC, we only need `InfoRefreshPeriod` if `Periodic` mode is specified.

File: config/manifests/bases/numaresources-operator.clusterserviceversion.yaml [hunk 3]
```diff
@@ -33,6 +33,11 @@ spec:
       - description: Group of Nodes to enable RTE on
         displayName: Group of nodes to enable RTE on
         path: nodeGroups
+      - description: Set the topology info refresh period. Use explicit 0 to disable.
+        displayName: Topology info refresh period setting
+        path: nodeGroups[0].config.infoRefreshPeriod
```
Comment (swatisehgal) [inline config/manifests/bases/numaresources-operator.clusterserviceversion.yaml:38]: Similar to my query above, would really appreciate some clarification here.

**Review by swatisehgal — COMMENTED** [review]
Thanks for the PR.
Looking good, just have a few minor queries.

File: bundle/manifests/numaresources-operator.clusterserviceversion.yaml (hunk 1 above)
Comment (fromanirh) [reply bundle/manifests/numaresources-operator.clusterserviceversion.yaml:68]: good catch, I don't have a answer, need to check.

File: pkg/objectstate/merge/merge.go (hunk 2 above)
Comment (fromanirh) [reply pkg/objectstate/merge/merge.go:114]: this needs clarification indeed. Thanks for pointing this out.

File: config/manifests/bases/numaresources-operator.clusterserviceversion.yaml (hunk 3 above)
Comment (fromanirh) [reply config/manifests/bases/numaresources-operator.clusterserviceversion.yaml:38]: agreed

File: pkg/objectstate/merge/merge.go (hunk 2 above)
Comment (fromanirh) [reply pkg/objectstate/merge/merge.go:114]: I think the least surprising behaviour is to propagate fields in the merge code, and make logic about the relationship between fields where we consume the data. I'm gonna change the logic which translate the settings into DS args, but keep the merge logic as simple as possible - hence updating all the fields when != nil.

File: bundle/manifests/numaresources-operator.clusterserviceversion.yaml (hunk 1 above)
Comment (fromanirh) [reply bundle/manifests/numaresources-operator.clusterserviceversion.yaml:68]: the root cause was a typo in the `operator-sdk` annotation:
```
-       //+operator-sdk:csv:customresourcedefinition:type=spec,displayName="Topology info mechanism setting",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:text"}
+       //+operator-sdk:csv:customresourcedefinitions:type=spec,displayName="Topology info mechanism setting",xDescriptors={"urn:alm:descriptor:com.tectonic.ui:text"}
```

**Comment by swatisehgal** [comment]
Thanks for addressing all the review comments.

/lgtm
/approve

**Pull request merged**